tsdb/blockwriter.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tsdb
+
+import (
+	"context"
+	"io/ioutil"
+	"math"
+	"os"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+
+	"github.com/prometheus/prometheus/pkg/timestamp"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+)
+
+// BlockWriter is a block writer that allows appending and flushing series to disk.
+type BlockWriter struct {
+	logger         log.Logger
+	destinationDir string
+
+	head      *Head
+	blockSize int64 // in ms
+	chunkDir  string
+}
+
+func durationToMillis(t time.Duration) int64 {
+	return int64(t.Seconds() * 1000)
+}
+
+// NewBlockWriter create a new block writer.
+//
+// The returned writer accumulates all the series in the Head block until `Flush` is called.
+//
+// Note that the writer will not check if the target directory exists or
+// contains anything at all. It is the caller's responsibility to
+// ensure that the resulting blocks do not overlap etc.
+// Writer ensures the block flush is atomic (via rename).
+func NewBlockWriter(logger log.Logger, dir string, blockSize int64) (*BlockWriter, error) {
+	w := &BlockWriter{
+		logger:         logger,
+		destinationDir: dir,
+		blockSize:      blockSize,
+	}
+	err := w.initHead()
+	if err != nil {
+		return nil, err
+	}
+	return w, nil
+}
+
+// initHead creates and initialises a new TSDB head.
+// blockSize should be in milliseconds.
```

Comment:
The tsdb doesn't know anything about the units it is used with, so this comment is incorrect.

Reply:
This comment is invalid now anyway and should be removed as `blockSize` is now passed in `NewBlockWriter` and does not mention milliseconds.